docs/RFCS/20210519_bounded_staleness_reads.md

```diff
@@ -0,0 +1,912 @@
+at the maximum timestamp that can be served locally and risk being wrong.
+
+The secondary benefit of bounded staleness reads over exact staleness reads is
+that they **minimize staleness** when possible. Whereas with exact staleness
+reads, which must conservatively pick a historical query timestamp in advance
+which provides a sufficiently high probability of being served locally, with
+bounded staleness reads, the query timestamp is chosen dynamically and the least
+stale timestamp that can be served locally is used.
+
+# Technical design
+
+The design of bounded staleness reads will be broadly similar to their design in
+Google's Spanner (see [prior art](#google-spanner)).
+
+Bounded staleness reads will be exposed through a pair of new functions that can
+be passed to an `AS OF SYSTEM TIME` clause:
+- `SELECT ... FROM ... AS OF SYSTEM TIME with_min_timestamp(TIMESTAMP)`
+- `SELECT ... FROM ... AS OF SYSTEM TIME with_max_staleness(INTERVAL)`
+
+`with_min_timestamp(TIMESTAMP)` defines a minimum timestamp to perform the
+bounded staleness read at. The actual timestamp of the read may be equal to or
+later than the provided timestamp, but can not be before the provided timestamp.
+This is useful to request a read from nearby followers, if possible, while
+enforcing causality between an operation at some point in time and any dependent
+reads.
+
+`with_max_staleness(INTERVAL)` defines a maximum staleness interval to perform
+the bounded staleness read with. The timestamp of the read may be less stale,
+but can be at most this stale with respect to the current time. This is useful
+to request a read from nearby followers, if possible, while placing some limit
+on how stale results can be.
+
+## Semantics
+
+### Limitations
+
+Bounded staleness reads will be limited to single-statement ("implicit"),
+read-only transactions. This means that explicit transactions, those surrounded
+by a `BEGIN` and `COMMIT` statement, can not use bounded staleness reads. The
+reason for this limitation is that the selection of the transaction timestamp
+requires knowledge of which rows will be read, so the timestamp selection cannot
+be performed before the SQL gateway has seen the entire transaction.
+
+Additionally, the initial implementations of bounded staleness reads will be
+limited to a subset of all read-only query plans. This limitation will be lifted
+incrementally over three distinct stages.
+- **Stage 1 - single-scan, single-range queries**
+- **Stage 2 - single-scan, multi-range queries**
+- **Stage 3 - multi-scan, multi-range queries**
+
+We define a **single-scan** query as one that contains only a single `ScanExpr`,
+and by extension, only a single `TableReader`.
+
+We define a **single-range** query as one that touches data on only a single KV
+range. Because range splits can be added between any two rows in a table in
+response to size and load factors, this effectively limits queries to those that
```

Comment:
This is just an implementational concern though, right? You're saying "because we're not going to really be able to check that all the things we're going to read are within a range we will be pessimistically assuming that all ranges are single-row". At that point it seems like stage 1 is really just single-row queries, and we should add single-range queries as a separate stage if we're even going to ever have them.